### PR TITLE
Simplifying our forklift testing pipeline definitions

### PR DESCRIPTION
**Pull request by evgeni**
this is based on the [Simplifying our forklift pipeline definitions RFC](https://community.theforeman.org/t/simplifying-our-forklift-pipeline-definitions/11692)

### CODE REVIEW

File: pipelines/pipeline_katello.yml [hunk 1]
```diff
@@ -64,21 +60,23 @@
...
     - foreman_installer
 
-- hosts: pipeline-katello-3.8-centos7
+- hosts: pipeline-katello-{{ katello_version }}-centos7
   become: true
   vars:
     bats_tests:
+      - fb-verify-packages.bats
       - fb-test-katello.bats
-      - fb-test-puppet.bats
```
Comment (ekohl) [inline pipelines/pipeline_katello.yml]: Why can we no longer test this?

File: pipelines/pipeline_katello.yml [hunk 2]
```diff
@@ -28,7 +28,7 @@
     - foreman_repositories
     - katello_repositories
     - role: foreman_client_repositories
-      when: katello_version in ['3.9', 'nightly']
+      when: katello_version is version('3.9', '>=')
```
Comment (ekohl) [inline pipelines/pipeline_katello.yml]: Is nightly considered newer than 3.9?

File: pipelines/pipeline_katello_upgrade.yml [hunk 3]
```diff
@@ -41,51 +45,57 @@
 - hosts: pipeline-upgrade-centos7
```
Comment (ekohl) [inline pipelines/pipeline_katello_upgrade.yml]: This step is now repeated a few times. I wonder if it makes sense to factor it out to a role you you can easily apply that a few times.

File: pipelines/pipeline_katello.yml (hunk 1 above)
Comment (evgeni) [reply pipelines/pipeline_katello.yml]: probably an oversight

File: pipelines/pipeline_katello.yml (hunk 2 above)
Comment (evgeni) [reply pipelines/pipeline_katello.yml]: yes (you acked PRs with that already :P https://github.com/theforeman/forklift/pull/877)

File: pipelines/pipeline_katello_upgrade.yml (hunk 3 above)
Comment (evgeni) [reply pipelines/pipeline_katello_upgrade.yml]: good idea

**Comment by evgeni** [comment]
todo/thoughts:
* rename `katello_version_N` in the upgrade pipeline to `katello_version_start`, `katello_version_intermediate` and `katello_version_final`, and make the intermediate one optional, so that we also support "one step" upgrade pipelines (as we currently have with 3.8, as 3.6 is gone)
* make the upgrade step DRY-er, as it repeats itself quite a bit

File: pipelines/pipeline_katello.yml [hunk 4]
```diff
@@ -1,31 +1,23 @@
...
-    forklift_name: pipeline-katello-nightly
-    forklift_boxes:
-      pipeline-katello-nightly-centos7:
-        box: centos7
-        memory: 8192
-      pipeline-proxy-nightly-centos7:
-        box: centos7
-        memory: 3072
+  vars_files:
+    - vars_katello_{{ katello_version }}.yml
```
Comment (ehelms) [inline pipelines/pipeline_katello.yml]: How about a `vars` directory instead of underscoring?

**Comment by ehelms** [comment]
I wonder if we should follow a similar naming convention to our jenkins jobs. Going specific to generic (e.g. katello_upgrade_pipeline, katello_pipeline)

File: pipelines/pipeline_katello_upgrade.yml [hunk 5]
```diff
@@ -25,6 +24,9 @@
       - "--puppet-server-jvm-max-heap-size 1G"
     foreman_installer_additional_packages:
       - katello
+  vars_files:
+    - vars_katello_upgrade_{{ katello_version }}.yml
+    - vars_katello_{{ katello_version_1 }}.yml
```
Comment (ehelms) [inline pipelines/pipeline_katello_upgrade.yml]: What is is `katello_version_1` ?

File: pipelines/vars_katello_upgrade_nightly.yml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+katello_version_1: '3.8'
+katello_version_2: '3.9'
+katello_version_3: '{{ katello_version }}'
```
Comment (ehelms) [inline pipelines/vars_katello_upgrade_nightly.yml]: Would it make sense to do `_N`, `_N-1`, `_N-2` ?

File: pipelines/pipeline_katello.yml (hunk 2 above)
Comment (evgeni) [reply pipelines/pipeline_katello.yml]: funnily, this only works on python2, so I will check for nightly too :/

**Comment by ehelms** [comment]
@evgeni Is `waiting on contributor` the correct status still?

**Comment by evgeni** [comment]
It's not ready to be merged just yet as I want to clean up the duplicated code, but I'd also love to hear more feedback.
So "waiting on contributor" + "waiting in reviewer"? :)

**Comment by ehelms** [comment]
Please provide examples of how these pipelines are now ran, e.g. I want to run the Katello nightly upgrade pipeline. These eventually need to go into `docs/testing.md`

**Comment by evgeni** [comment]
@ehelms documented, also added a block how to add new versions

**Comment by evgeni** [comment]
OK, this is as DRY as I can do it w/o creating new meta roles, which I'd like to postpone for now.

File: docs/testing.md [hunk 7]
```diff
@@ -48,11 +48,53 @@ Or if you want to run bats from a different repository or branch, edit `centos7-
...
+* `katello_pipeline` - Installs a Katello and a Content Proxy VMs and runs the `foreman_testing` role to verify the setup.
+  Expects the `katello_version` variable to be set to a known version (currently: 3.8, 3.9, 3.10, nightly)
+* `katello_upgrade_pipeline` - Installs a Katello VM, upgrades it twice and runs the `foreman_testing` role to verify the final upgrade.
+  Expects the `katello_version` variable to be set to a known version (currently: 3.8, 3.9, 3.10, nightly).
+* `pipeline_foreman_nightly` - Installs a nightly Foreman VM and runs the `foreman_testing` role to verify the setup.
+
+#### Examples
+
+    ansible-playbook pipelines/katello_upgrade_pipeline.yml -e forklift_state=up -e katello_version=nightly
+    ansible-playbook pipelines/katello_pipeline.yml -e forklift_state=up -e katello_version=3.10
```
Comment (ehelms) [inline docs/testing.md]: Can do in a follow up, I've been thinking about making `up` default. Thoughts?

File: pipelines/vars/katello_upgrade_nightly.yml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+katello_version_start: '3.8'
+katello_version_intermediate: '3.9'
```
Comment (ehelms) [inline pipelines/vars/katello_upgrade_nightly.yml]: Mind updating to 3.9, 3.10 ?

Comment (evgeni) [reply pipelines/vars/katello_upgrade_nightly.yml]: done

File: docs/testing.md (hunk 7 above)
Comment (evgeni) [reply docs/testing.md]: Sounds good, but doesn't belong in here ;)

**Comment by evgeni** [comment]
will merge tomorrow so I can do a manual run afterwards :)

**Pull request merged**